### ISSUE TITLE
`PTH*`: Incorrect suggestion to use `Pathlib` when using file descriptors or byte strings

### ISSUE THREAD

**Issue by MichaReiser** [astral-sh/ruff#17699]
Multiple pathlib rules recommend the use of `pathlib.Path` over their `os` equivalent even when they're used with a bytes string or directory descriptor which aren't supported by pathlib. 

One such example is

```py
import os
os.listdir(1)
```

```
$ ruff --version
ruff 0.11.7
$ ruff check --isolated --select PTH208 t.py
t.py:2:1: PTH208 Use `pathlib.Path.iterdir()` instead.
  |
1 | import os
2 | os.listdir(1)
  | ^^^^^^^^^^ PTH208
  |

Found 1 error.
```

Original reported issues (see for more examples):

* [ ] `PTH123` https://github.com/astral-sh/ruff/issues/17691 (only sometimes)
* [ ] `PTH116`: https://github.com/astral-sh/ruff/issues/17693
* [ ] `PTH104`: https://github.com/astral-sh/ruff/issues/17694
* [ ] `PTH208`: https://github.com/astral-sh/ruff/issues/17695

Past instances of the same or similar error:
* https://github.com/astral-sh/ruff/issues/7620
* https://github.com/astral-sh/ruff/issues/12871

**Comment by LaBatata101** [comment]
I can take all of those, but if anyone wants to take any of the others, feel free to.  I'll start with #17691.

**Comment by sbrudenell** [comment]
I wrote the linked sub-issues.

IMO file descriptor usage shouldn't be conflated with `bytes` usage.

It's still valid to suggest `pathlib.Path` in place of a non-`pathlib` function used with a `bytes` path.

It's not the case that `pathlib` somehow supports `str` and doesn't support `bytes`. It actually just expects its arguments to be `fsdecode()`'ed strings.

That's to say, IMO the following *should* error:
```python
open(b"foo")  # *should* error and suggest Path(fsdecode(b"foo")).open()
```

In #17705 I see that `open(b"foo")` is added as a test.

This is a philosophical stance. In my experience, explicitly calling `fsdecode()` makes it clear where possible platform path encoding issues might arise.

File descriptors are a different access pattern that just doesn't have an analog in `pathlib`.

**Comment by ntBre** [comment]
Ah I see, I just tested `Path(b"foo")` and concluded that it should be included here because it also causes an error.

It does seem reasonable to me still to flag the bytes case if we update the rule docs with an example using `fsdecode`, but I'm curious what @MichaReiser and @LaBatata101 think too.

Thanks for opening the issues and following up here!

**Comment by LaBatata101** [comment]
I think that's reasonable, we could even provide a fix if there isn't already.

**Comment by MichaReiser** [comment]
Thanks @sbrudenell for the added context. I only read the pathlib documentation and it stated that byte strings aren't supported. 

Let's only change the file descriptor case for now. We need to be more careful about byte strings if we start providing fixes (although they have to be unsafe in most cases anyway because we can't tell in many instances if a path is a file descriptor, string or byte string)

**Comment by ntBre** [comment]
I had forgotten about this issue, and in the meantime we added preview fixes for nearly all of the PTH rules. We need to go back through these fixes and make almost all of them almost always unsafe, or at least update the fix safety docs for those that are already always unsafe, before we can stabilize them.

**Comment by alzeph** [comment]
Went through the remaining `PTH*` rules to see what's left here regarding file descriptors specifically (as opposed to the bytes-path side of this, which #24701 is tracking separately).

Opened #27826 for `os.path.getsize`/`getatime`/`getmtime`/`getctime`/`isdir`/`isfile`/`exists`/`samefile`. These are pure-Python wrappers around `os.stat`, so they transitively accept a file descriptor at runtime even though that's not part of their documented signature, and none of them had the `is_file_descriptor` guard that `os.chmod`/`os.stat`/`os.listdir`/`open` already have.

I also checked `os.rename`, `os.remove`, `os.unlink`, `os.rmdir`, `os.replace`, `os.readlink`, and `os.symlink` since they're flagged by PTH104/105/106/107/108/115/211 and none of them had a guard either. Turns out none of these actually accept a file descriptor for their path argument in the first place:

```python
>>> os.rename(1, "b")
TypeError: rename: src should be string, bytes or os.PathLike, not int
>>> os.remove(1)
TypeError: remove: path should be string, bytes or os.PathLike, not int
```

`os.supports_fd` confirms this: `os.rename`/`remove`/`unlink`/`rmdir`/`replace`/`readlink`/`symlink` aren't in it. So passing an int there is already broken regardless of the pathlib suggestion, meaning there's no real false positive to fix for those seven rules. Leaving this note here in case anyone else goes looking.

`os.path.islink` also stays as-is: it uses `os.lstat`, which doesn't support fd either.

So after #27826, the file-descriptor side of this issue should be fully covered for the rules that actually support it.
